File: P09/E1/monitorImpresion.java
/**
* P9 - E1
* Clase monitorImpresion, compuesta por un contructor del monitor, un método pedir_imp y otro dejar_imp
* @author Carlos Gallardo Polanco
* @version 18/12/2017
*/

public class monitorImpresion{
	private int  i=0;
	private static int impresoras=3;
	private static boolean[] estadoimpresora = new boolean[3];

	/**
    * Contructor de la clase monitorImpresion
    */
	public monitorImpresion(){
		for(i=0;i<3;i++){
			estadoimpresora[i]=true;
		}
	}

	/**
    *Metodo pedir_imp de la clase monitorImpresion
    */
	public synchronized int pedir_imp(){
		while(impresoras==0){
			try {
         		wait();
      		} catch (InterruptedException e) {}
		}
		int n=0;
		while(!estadoimpresora[n]){n++;}
		estadoimpresora[n]=false;
		impresoras--;
		notifyAll();
		return n;
	}

	/**
    *Metodo dejar_imp de la clase monitorImpresion
    */
	public synchronized int dejar_imp(){
		while(impresoras==3){
			try {
         		wait();
      		} catch (InterruptedException e) {}
		}
		int n=0;
		while(estadoimpresora[n]){n++;}
		estadoimpresora[n]=true;
		impresoras++;
		notifyAll();
		return n;
	}
}

File: P09/E1/monitorImpresion.pas
Monitor monitorimpresion
	var
		i, impresoras: integer;
		estadoimpresora: array [1..3] of Boolean;
		imp: condition;

	procedure pedir_imp (var n: integer)
		begin
			if impresoras==0 then wait(imp);
			n:=1;
			while(!estadoimpresora[n]) do n:=n+1;
			estadoimpresora[n]:=false;
			impresoras--;
			send(imp);
		end

	procedure dejar_imp (var n: integer)
		begin
      if impresoras==3 then wait(imp);
			n:=1;
			while(estadoimpresora[n]) do n:=n+1;
			estadoimpresora[n]:=true;
			impresoras++;
			send(imp);
		end

	begin (*inicialización*)
		for i=1 to 3 do
			estadoimpresora[i]:=true;
		impresoras:=3;
	end
